File: ganapathy_pavers/ganapathy_pavers/report/internal_material_transfer/internal_material_transfer.js
// Copyright (c) 2023, Thirvusoft and contributors
// For license information, please see license.txt
/* eslint-disable */

var vehicles = [], drivers = [], operators = [];

let refresh_fn = frappe.query_report.refresh
frappe.query_report.refresh = async (...args) => {
	vehicles = await frappe.db.get_list("Stock Entry", {
		limit: 0,
		filters: {
			docstatus: 1,
			is_internal_stock_transfer: 1,
			stock_entry_type: 'Material Transfer'
		},
		fields: ['vehicle'],
		pluck: 'vehicle'
	})

	drivers = await frappe.db.get_list("Stock Entry", {
		limit: 0,
		filters: {
			docstatus: 1,
			is_internal_stock_transfer: 1,
			stock_entry_type: 'Material Transfer'
		},
		fields: ['driver'],
		pluck: 'driver'
	})

	operators = await frappe.db.get_list("Stock Entry", {
		limit: 0,
		filters: {
			docstatus: 1,
			is_internal_stock_transfer: 1,
			stock_entry_type: 'Material Transfer'
		},
		fields: ['operator'],
		pluck: 'operator'
	})
	refresh_fn.call(frappe.query_report, ...args)
}


frappe.query_reports["Internal Material Transfer"] = {
	filters: [
		{
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			reqd: 1,
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -7)
		},
		{
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			reqd: 1,
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -1)
		},
		{
			fieldname: 'item_code',
			label: __('Item'),
			fieldtype: 'MultiSelectList',
			get_data: async function (txt) {
				let filters = [
					['Stock Entry', 'docstatus', '=', '1'],
					['Stock Entry', 'stock_entry_type', '=', 'Material Transfer'],
					['Stock Entry', 'is_internal_stock_transfer', '=', '1']
				];

				if (frappe.query_report.get_filter_value('from_date')) {
					filters.push(['Stock Entry', 'posting_date', '>=', frappe.query_report.get_filter_value('from_date')])
				}
				if (frappe.query_report.get_filter_value('to_date')) {
					filters.push(['Stock Entry', 'posting_date', '<=', frappe.query_report.get_filter_value('to_date')])
				}
				if (frappe.query_report.get_filter_value('s_warehouse')?.length) {
					filters.push(['Stock Entry Detail', 's_warehouse', 'in', frappe.query_report.get_filter_value('s_warehouse')])
				}
				if (frappe.query_report.get_filter_value('t_warehouse')?.length) {
					filters.push(['Stock Entry Detail', 't_warehouse', 'in', frappe.query_report.get_filter_value('t_warehouse')])
				}
				if (frappe.query_report.get_filter_value('vehicle')) {
					filters.push(['Stock Entry', 'vehicle', '=', frappe.query_report.get_filter_value('vehicle')])
				}
				if (frappe.query_report.get_filter_value('driver')) {
					filters.push(['Stock Entry', 'driver', '=', frappe.query_report.get_filter_value('driver')])
				}
				if (frappe.query_report.get_filter_value('operator')) {
					filters.push(['Stock Entry', 'operator', '=', frappe.query_report.get_filter_value('operator')])
				}

				return frappe.db.get_list("Stock Entry", {
					filters: filters,
					group_by: "item_code",
					fields: ['`tabStock Entry Detail`.item_code as value', '`tabStock Entry Detail`.item_name as description'],
					limit: 0
				});
			}
		},
		{
			fieldname: 's_warehouse',
			label: __('From Warehouse'),
			fieldtype: 'MultiSelectList',
			get_data: function (txt) {
				let filters = [
					['Stock Entry', 'docstatus', '=', '1'],
					['Stock Entry', 'stock_entry_type', '=', 'Material Transfer'],
					['Stock Entry', 'is_internal_stock_transfer', '=', '1']
				];

				if (frappe.query_report.get_filter_value('from_date')) {
					filters.push(['Stock Entry', 'posting_date', '>=', frappe.query_report.get_filter_value('from_date')])
				}
				if (frappe.query_report.get_filter_value('to_date')) {
					filters.push(['Stock Entry', 'posting_date', '<=', frappe.query_report.get_filter_value('to_date')])
				}
				if (frappe.query_report.get_filter_value('item_code')?.length) {
					filters.push(['Stock Entry Detail', 'item_code', 'in', frappe.query_report.get_filter_value('item_code')])
				}
				if (frappe.query_report.get_filter_value('t_warehouse')?.length) {
					filters.push(['Stock Entry Detail', 't_warehouse', 'in', frappe.query_report.get_filter_value('t_warehouse')])
				}
				if (frappe.query_report.get_filter_value('vehicle')) {
					filters.push(['Stock Entry', 'vehicle', '=', frappe.query_report.get_filter_value('vehicle')])
				}
				if (frappe.query_report.get_filter_value('driver')) {
					filters.push(['Stock Entry', 'driver', '=', frappe.query_report.get_filter_value('driver')])
				}
				if (frappe.query_report.get_filter_value('operator')) {
					filters.push(['Stock Entry', 'operator', '=', frappe.query_report.get_filter_value('operator')])
				}

				return frappe.db.get_list("Stock Entry", {
					filters: filters,
					group_by: "s_warehouse",
					fields: ['`tabStock Entry Detail`.s_warehouse as value', '`tabStock Entry Detail`.s_warehouse as description'],
					limit: 0
				});
			}
		},
		{
			fieldname: 't_warehouse',
			label: __('To Warehouse'),
			fieldtype: 'MultiSelectList',
			get_data: function (txt) {
				let filters = [
					['Stock Entry', 'docstatus', '=', '1'],
					['Stock Entry', 'stock_entry_type', '=', 'Material Transfer'],
					['Stock Entry', 'is_internal_stock_transfer', '=', '1']
				];

				if (frappe.query_report.get_filter_value('from_date')) {
					filters.push(['Stock Entry', 'posting_date', '>=', frappe.query_report.get_filter_value('from_date')])
				}
				if (frappe.query_report.get_filter_value('to_date')) {
					filters.push(['Stock Entry', 'posting_date', '<=', frappe.query_report.get_filter_value('to_date')])
				}
				if (frappe.query_report.get_filter_value('item_code')?.length) {
					filters.push(['Stock Entry Detail', 'item_code', 'in', frappe.query_report.get_filter_value('item_code')])
				}
				if (frappe.query_report.get_filter_value('s_warehouse')?.length) {
					filters.push(['Stock Entry Detail', 's_warehouse', 'in', frappe.query_report.get_filter_value('s_warehouse')])
				}
				if (frappe.query_report.get_filter_value('vehicle')) {
					filters.push(['Stock Entry', 'vehicle', '=', frappe.query_report.get_filter_value('vehicle')])
				}
				if (frappe.query_report.get_filter_value('driver')) {
					filters.push(['Stock Entry', 'driver', '=', frappe.query_report.get_filter_value('driver')])
				}
				if (frappe.query_report.get_filter_value('operator')) {
					filters.push(['Stock Entry', 'operator', '=', frappe.query_report.get_filter_value('operator')])
				}

				return frappe.db.get_list("Stock Entry", {
					filters: filters,
					group_by: "t_warehouse",
					fields: ['`tabStock Entry Detail`.t_warehouse as value', '`tabStock Entry Detail`.t_warehouse as description'],
					limit: 0
				});
			}
		},
		{
			fieldname: 'vehicle',
			label: __('Vehicle'),
			fieldtype: 'Link',
			options: 'Vehicle',
			get_query: () => {
				return {
					filters: {
						name: ["in", vehicles]
					}
				}
			}
		},
		{
			fieldname: 'driver',
			label: __('Driver'),
			fieldtype: 'Link',
			options: 'Driver',
			get_query: () => {
				return {
					filters: {
						name: ["in", drivers]
					}
				}
			}
		},
		{
			fieldname: 'operator',
			label: __('Operator'),
			fieldtype: 'Link',
			options: 'Driver',
			get_query: () => {
				return {
					filters: {
						name: ["in", operators]
					}
				}
			}
		},
		{
			fieldname: 'group_by',
			label: __('Group By'),
			fieldtype: 'MultiSelectList',
			get_data: function (txt) {
				return [
					{
						value: 'Date',
						description: ''
					},
					{
						value: 'Item',
						description: ''
					},
				]
			},
			on_change: () => {
				if (frappe.query_report.get_filter_value('report_type') == 'Summary' && !frappe.query_report.get_filter_value('group_by')?.length) {
					frappe.show_alert({
						message: '<b>Summary</b> will be generated based on the <b>Group By<b>',
						indicator: 'yellow'
					});
				}
				frappe.query_report.refresh();
			}
		},
		{
			fieldname: 'also_group_by',
			label: __('Also Group By'),
			fieldtype: 'MultiSelectList',
			get_data: function (txt) {
				return [
					{
						value: 'Vehicle',
						description: ''
					},
					{
						value: 'Driver',
						description: ''
					},
					{
						value: 'Operator',
						description: ''
					}
				];
			},
			depends_on: "eval: frappe.query_report.get_filter_value('group_by')?.length"
		},
		{
			fieldname: 'report_type',
			label: __('Report Type'),
			fieldtype: 'Select',
			options: 'Report\nSummary',
			default: 'Report',
			on_change: () => {
				if (frappe.query_report.get_filter_value('report_type') == 'Summary' && !frappe.query_report.get_filter_value('group_by')?.length) {
					frappe.show_alert({
						message: '<b>Summary</b> will be generated based on the <b>Group By<b>',
						indicator: 'yellow'
					});
				}
				frappe.query_report.refresh();
			}
		},
	],
	formatter: function (value, row, column, data, default_formatter) {
		value = __(default_formatter(value, row, column, data));
		if (data && ["Group Total", "Total"].includes(data.item_code)) {
			value = `<b>${value}</b>`
		}

		if (column.fieldname == 'posting_date' && ['Group Total', 'Total'].includes(data.item_code)) {
			value = `<b>${data.item_code}</b>`
		}

		return value
	}
};

let setup_filters_fn = frappe.query_report.setup_filters;

frappe.query_report.setup_filters = (...args) => {
	setup_filters_fn.call(frappe.query_report, ...args)
	frappe.query_report.set_filter_value('group_by', ['Date']);
}

File: ganapathy_pavers/ganapathy_pavers/report/internal_material_transfer/internal_material_transfer.py
# Copyright (c) 2023, Thirvusoft and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
	for field in ['s_warehouse', 't_warehouse', 'item_code', 'group_by', 'also_group_by']:
		if not filters.get(field):
			filters[field] = []

	columns, data = get_columns(filters), get_data(filters)
	return columns, data

def get_data(filters):
	order_by = ', '.join([{
		'Date': 'se.posting_date',
		'Item': 'sed.item_code'
	}.get(grp) or '' for grp in filters.get('group_by') or []])
	
	group_by=''

	if filters.get('report_type') == 'Summary':
		group_by = order_by

	if filters.get('group_by'):
		also_group_by = ", ".join([f"se.{frappe.scrub(field)}" for field in (filters.get('also_group_by') or [])])
	else:
		also_group_by = ""

	conditions = ''

	if filters.get('from_date'):
		conditions += f""" and se.posting_date >= DATE('{filters.get('from_date')}') """
	
	if filters.get('to_date'):
		conditions += f""" and se.posting_date <= DATE('{filters.get('to_date')}') """
	
	if filters.get('item_code'):
		conditions += f""" and sed.item_code in ({', '.join([f" '{i}' " for i in filters.get('item_code')])}) """

	if filters.get('s_warehouse'):
		conditions += f""" and sed.s_warehouse in ({', '.join([f" '{i}' " for i in filters.get('s_warehouse')])}) """
	
	if filters.get('t_warehouse'):
		conditions += f""" and sed.t_warehouse in ({', '.join([f" '{i}' " for i in filters.get('t_warehouse')])}) """
	
	if filters.get('vehicle'):
		conditions += f""" and se.vehicle = '{filters.get('vehicle')}' """

	if filters.get('driver'):
		conditions += f""" and se.driver = '{filters.get('driver')}' """
	
	if filters.get('operator'):
		conditions += f""" and se.operator = '{filters.get('operator')}' """


	query = f"""
		SELECT
			se.name as docname,
			se.posting_date,
			sed.item_code,
			SUM(sed.qty) as qty,
			sed.uom,
			SUM(sed.qty*ifnull(
				(
					SELECT
						uom.conversion_factor
					FROM `tabUOM Conversion Detail` uom
					WHERE
						uom.parenttype='Item' and
						uom.parent=sed.item_code and
						uom.uom=sed.uom
				)
				, 0)/
				ifnull((
					SELECT
						uom.conversion_factor
					FROM `tabUOM Conversion Detail` uom
					WHERE
						uom.parenttype='Item' and
						uom.parent=sed.item_code and
						uom.uom='SQF'
				)    
				, 0)
			) as sqr_ft,
			SUM(IFNULL(sed.ts_qty, 0)) as bdl,
			SUM(IFNULL(sed.pieces, 0)) as pieces,
			se.vehicle as vehicle,
			se.driver as driver,
			se.operator as operator
		FROM `tabStock Entry Detail` sed
		INNER JOIN `tabStock Entry` se
		ON se.name = sed.parent
		WHERE
			se.docstatus = 1 and
			se.stock_entry_type = 'Material Transfer' and
			se.is_internal_stock_transfer = 1
			{conditions}
		GROUP BY {'sed.name,' if not group_by else ''} sed.uom {', '+ group_by if group_by else ''} {', '+also_group_by if also_group_by else ''}
		{
			f'''ORDER BY 
			{order_by} {', '+also_group_by if order_by and also_group_by else also_group_by}'''
			if order_by or also_group_by
			else f'''
				ORDER BY se.posting_date, sed.item_code
			'''
		}
	"""

	data = frappe.db.sql(query=query, as_dict=True)

	if filters.get('report_type') == "Report":
		if filters.get('group_by'):
			data = add_group_total(data=data, filters=filters, add_total_row_only = False)
		else:
			data = add_group_total(data=data, filters=filters, add_total_row_only = True)
			frappe.msgprint('Please choose <b>Group By</b> to get a <b>Group Total</b>', alert=True, indicator='yellow')
	else:
		data = add_group_total(data=data, filters=filters, add_total_row_only = True)

	return data

def add_group_total(data, filters, gt_column = "item_code", add_total_row_only=False):
	res = []
	if filters.get('group_by') and filters.get('also_group_by'):
		filters['group_by'] += filters.get('also_group_by')

	current_group = "---".join([str(data[0].get({'Date': 'posting_date', 'Item': 'item_code'}.get(gb) or frappe.scrub(gb))) for gb in filters.get('group_by')]) if data else False
	group=[]
	group_total = {}
	total = {}

	for row in data:
		groupby = "---".join([str(row.get({'Date': 'posting_date', 'Item': 'item_code'}.get(gb) or frappe.scrub(gb))) for gb in filters.get('group_by')])
		if group and groupby != current_group and not add_total_row_only:
			group.sort(key = lambda x: (str(x.get('posting_date') or ''), str(x.get('item_code') or '')))
			res += group
			group_total[gt_column] = "Group Total"
			res.append(group_total)
			res.append({})
			group=[]
			group_total = {} 
			current_group = groupby

		group.append(row)
		for col in row:
			if (col not in group_total and (isinstance(row[col], int) or isinstance(row[col], float))):
				group_total[col] = (group_total.get(col) or 0) + row[col]
				total[col] = (total.get(col) or 0) + row[col]
				
			elif (isinstance(row[col], int) or isinstance(row[col], float)):
				group_total[col] = (group_total.get(col) or 0) + row[col]
				total[col] = (total.get(col) or 0) + row[col]

	if group and not add_total_row_only:
		group.sort(key = lambda x: (str(x.get('posting_date') or ''), str(x.get('item_code') or '')))
		res += group
		group_total[gt_column] = "Group Total"
		res.append(group_total)
		group=[]
		group_total = {} 
		groupby = "---".join([str(row.get({'Date': 'posting_date', 'Item': 'item_code'}.get(gb) or frappe.scrub(gb))) for gb in filters.get('group_by')])
		current_group = groupby
	
	if not res and group:
		res = group
	if res:
		total[gt_column] = "Total"
		res.append(total)

	return res

def get_columns(filters):
	columns = [
		{
			'fieldname': 'docname',
			'label': 'Document No.',
			'fieldtype': 'Link',
			'options': 'Stock Entry',
			'width': 150,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary')
		},
		{
			'fieldname': 'posting_date',
			'label': 'Date',
			'fieldtype': 'Date',
			'width': 150,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary' and 'Date' not in (filters.get('group_by') or []))
		},
		{
			'fieldname': 'item_code',
			'label': 'Item',
			'fieldtype': 'Link',
			'options': 'Item',
			'width': 350,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary' and 'Item' not in (filters.get('group_by') or []))
		},
		{
			'fieldname': 'qty',
			'label': 'Qty',
			'fieldtype': 'Float',
			'width': 150,
		},
		{
			'fieldname': 'uom',
			'label': 'UOM',
			'fieldtype': 'Link',
			'options': 'UOM',
			'width': 100,
		},
		{
			'fieldname': 'sqr_ft',
			'label': 'SQF',
			'fieldtype': 'Float',
			'width': 150,
		},
		{
			'fieldname': 'bdl',
			'label': 'Bdl',
			'fieldtype': 'Float',
			'width': 150,
		},
		{
			'fieldname': 'pieces',
			'label': 'Pieces',
			'fieldtype': 'Float',
			'width': 150,
		},
		{
			'fieldname': 'vehicle',
			'label': 'Vehicle',
			'fieldtype': 'Link',
			'options': 'Vehicle',
			'width': 150,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary' and 'Vehicle' not in (filters.get('also_group_by') or []))
		},
		{
			'fieldname': 'driver',
			'label': 'Driver',
			'fieldtype': 'Link',
			'options': 'Driver',
			'width': 150,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary' and 'Driver' not in (filters.get('also_group_by') or []))
		},
		{
			'fieldname': 'operator',
			'label': 'Operator',
			'fieldtype': 'Link',
			'options': 'Driver',
			'width': 150,
			'hidden': bool(filters.get('group_by') and filters.get('report_type') == 'Summary' and 'Operator' not in (filters.get('also_group_by') or []))
		}
	]

	return columns
